raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -2250,7 +2249,40 @@ private Long append(int epoch, List<T> records, boolean isAtomic) {
 
     @Override
     public void resign(int epoch) {
-        throw new UnsupportedOperationException();
+        if (epoch < 0) {
+            throw new IllegalArgumentException("Attempt to resign from an invalid negative epoch " + epoch);
+        }
+
+        if (!quorum.isVoter()) {
+            throw new IllegalArgumentException("Attempt to resign by a non-voter");
+        }
+
+        LeaderAndEpoch leaderAndEpoch = leaderAndEpoch();
+        int currentEpoch = leaderAndEpoch.epoch();
+
+        if (epoch > currentEpoch) {
+            throw new IllegalArgumentException("Attempt to resign from epoch " + epoch +
+                " which is larger than the current epoch " + currentEpoch);
+        } else if (epoch < currentEpoch) {
+            // If the passed epoch is smaller than the current epoch, then it might mean
+            // that the listener has not been notified about a leader change that already
+            // took place. In this case, we consider the call as already fulfilled and
+            // take no further action.
+            return;
+        } else if (!leaderAndEpoch.isLeader(quorum.localIdOrThrow())) {
```

Comment:
Does it make sense to do this check before the epoch validation? If we're not the leader and received an old epoch (which, if i understand, seems likely if we're _not_ the leader anymore), we will silently ignore in the above case.

Reply:
If the epoch has moved on, then the leader check is likely to fail, so the current order seems to make sense. We don't keep a history of previous states, so I think the best we can do is catch cases where the passed epoch does not make sense with the current state.

Reply:
I see what you mean, and yea that is a fair point 👍